Clean up Markdown generation from witx

kubkon opened the pull request:
This commit tweaks `docs.rs` module of the `witx` tool facilitating
generation of Markdown documentation from `*.witx` files which
* is formatted in a more readable way
* inserts HTML links for each header but also a list entry (in the
  latter case, the link is scoped; i.e., it includes the link id
  of the enclosing Markdown section)

---
**TODO (in no particular order):**
- [x] document introduced helper types
~~- [ ] add sanity tests~~
- [x] try generating a link, and revert if dead + log a warning

---
EDIT: More updates in the comments below. Have a look!

kubkon commented:
While we're here, I thought I'm going to raise an issue of how we should specify/tag links in `*.witx` files. Firstly, it seems we use backticks in comment blocks for either rendering the word as a type, e.g., `u32`, or what it looks to me like trying to specify an internal link such as `fdflags::sync` or whatnot. I'd like to propose wrapping a potential link inside square brackets like it's normally done in Markdown; so we'd end up with something as follows:

```
[`fdflags::sync`]
```

Secondly, is there any particular reason why we use "::" for scoping? Could we not use the traditional Markdown/HTML "."? With the former, if "::" is left after expanding the link to full Markdown notation `[fdflags::sync](#fdflags::sync)` doesn't seem to properly reference anything tagged with `href="#fdflags::sync"` whereas if swapped for ".", works as expected.

Thirdly, and this is mainly targeting the case of multiple separate modules that saw the light of day in current state-of-the-art ephemeral release, we'll need to figure out how to handle inter-doc links. For instance, `wasi_ephemeral_args` module could reference something in `wasi_ephemeral_path` module, etc. Perhaps we could use Rust's approach of specifying the links manually, inside comment blocks directly in `*.witx`?

cc @pchickey @sunfishcode

sunfishcode commented:
Adding `[` and `]` around linkable things sounds reasonable to me.

No strong reason for `::`. I think that just emerged out of habit from folks familiar with Rust and C++, but `.` works for me.

We process all the witx files at once, so it seems like we should theoretically be able to resolve cross-file links, because we have all the links and their targets in memory at once. But if that's complicated, possibly a simpler approach would be to just write \[`path.open`\] to \[`path.open`\]\(#wasi_ephermeral_path.md#path.open\) just based on the filename. How does that sound?

pchickey commented:
Thanks so much for working on this, it looks excellent so far.

No strong opinion on `::` vs `.` as a scope separator. When I add scope to witx I'll likely follow your lead. My proposal in https://github.com/WebAssembly/WASI/issues/157 was to use `:`. Paint the shed any color you like :)

re bracketing - Other doc systems try to treat the contents of any `single_quoted` token as potentially linkable and insert a link if they find a destination. Some even warn if they fail to find a destination. If that is easy enough to do, might be nice.

kubkon commented:
In 3cafe13 I've introduced hopefully the final iteration of the approach at the problem. In summary:

This approach is heavily inspired by `HostRef` and related present in `wasmtime` (see [wasmtime/crates/api/ref.rs]). In this new approach, instead of holding all possible type variants inside a statically-typed `enum`, I've decided to have a tiered structure where (from bottom/inner-most element to the top/outer-most element):
* `Box<dyn MdElement>` is any "content-box" holding for instance a Markdown element representing a section, `NamedType`, etc.
* `MdNode` struct wrapping the "content-box" plus storing all children `MdNodes` (representing subsections, etc.), and a weak link to the parent `MdNode` (if any)
* `MdNodeRef` struct wrapping `MdNode` inside an `Rc<RefCell<_>>`, and offering convenience methods to extract `dyn MdElement` directly from `MdNodeRef`, etc.

Finally, `trait MdElement` represents the minimum interface required for any "content-box" to have in order to be able to interface with other generated Markdown nodes.

[wasmtime/crates/api/ref.rs]: https://github.com/bytecodealliance/wasmtime/blob/master/crates/api/src/ref.rs

Now that I've got the tree generation mostly out of the way (I hope!), I'll briefly explore what @pchickey suggested when it comes to link generation, namely, upon encountering a dead link, `witx` will *not* expand it and log a warning instead.

kubkon commented:
Right, I think this PR is ready for review. I've now tweaked the generator to try and render any `content` as linkable, and if the link doesn't exist, log a warning and preserve the original content. I've also added some high level docs to help in navigating and understanding the code a little bit better. In terms of tests, in order not to thwart the potential usefulness of this, I've decided to postpone adding them in a subsequent PR (this PR has been slowly growing becoming bigger and bigger).

Anyhow, example output of running `witx docs` against current `snapshot1` can be found in [kubkon/squishy-squashy] repo. Also, in order to log the warnings, I've introduced the simplest `pretty_env_logger::init()` call, so that, when generating docs from `*.witx`, you could do:

```
RUST_LOG=warn witx phases/snapshot/witx/wasi_snapshot_preview1.witx docs
```

yielding warnings:

```
...
WARN  witx::docs > Link [`fdflag::dsync`](#fdflag.dsync) could not be found in the document!
...
```

And so on.

[kubkon/squishy-squashy]: https://github.com/kubkon/squishy-squashy

pchickey commented on tools/witx/src/docs/ast.rs:
```diff
@@ -0,0 +1,487 @@
+    }
+}
+
+// TODO
```
this TODO is done right?

kubkon replied:
No, not yet. I purposely didn't touch the polyfill docs just yet, as I wanted to focus on the main docs first, and do the polyfill in a subsequent PR. Does that sound OK?

pchickey replied:
My bad I misread what the todo meant. I see now. Its fine to wait on the polyfill docs, they're not very important.

kubkon commented on tools/witx/src/docs/ast.rs:
```diff
@@ -0,0 +1,487 @@
+                name.to_owned()
+            };
+            let tt = MdNamedType::new("-", id.as_str(), name, &r#const.docs);
+            // TODO handle r#const.value
```
Sounds good. I can add such a field but I'd like to keep a `TODO` there to remind ourselves to verify whether a better approach could potentially be found.

kubkon replied:
On second thought, could we leave this one out until I add some tests and rethink `MdType` to be more recursive in nature? Otherwise, I worry this would not be correct. What do you think?

pchickey replied:
Works for me

kubkon commented on tools/witx/src/docs/ast.rs:
```diff
@@ -0,0 +1,487 @@
+use super::md::*;
+use super::Documentation;
+use crate::ast::*;
+use crate::polyfill::*;
```
Done in a1caed39.

kubkon commented on tools/witx/src/docs/ast.rs:
```diff
@@ -0,0 +1,487 @@
+
+impl ToMarkdown for Document {
+    fn generate(&self, node: MdNodeRef) {
+        let header = "#".repeat(node.borrow().ancestors().len() + 1);
```
Done in a0bacb1.
